xof.go

```diff
@@ -0,0 +1,50 @@
+package kyber
+
+import (
+	"crypto/cipher"
+	"io"
+)
+
+// An XOF is an extendable output function, which is a cryptographic
+// primitive that can take arbitrary input in the same way a hash
+// function does, and then create a stream of output, up to a limit
+// determined by the size of the internal state of the hash function
+// the underlies the XOF.
+type XOF interface {
+	// Write absorbs more data into the hash's state. It panics if called
+	// after Read. Use Reseed() to reset the XOF into a state where more data
+	// can be absorbed via Write.
+	io.Writer
+
+	// Read reads more output from the hash. It returns io.EOF if the
+	// limit of available data for reading has been reached.
+	io.Reader
+
+	// An XOF implements cipher.Stream, so that callers can use XORKeyStream
+	// to encrypt/decrypt data. The key stream is read from the XOF using
+	// the io.Reader interface.
+	cipher.Stream
+
+	// KeySize is the number of bytes that must be written to the XOF before
+	// it is ready to give secure output.
+	KeySize() int
```

Comment:
I'm very surprised about this. As I understood it, and from Bryan's comment, a XOF is supposed to be kinda of "superset" of a hash function. A Hash function, even fed with one byte outputs a "safe" output in the sense that it's looking random. Why does a XOF do not have this property ? Furthermore, I can't find any restrictions of the kind in  https://godoc.org/golang.org/x/crypto/blake2b for example.

Reply:
For blake, I set KeySize to blake.BlockSize, but I can't remember why. I need advice from a crypto expert on this, because I am having trouble sorting out what Cipher did from what I guessed myself while learning this...

It seems to me that the caller needs some way to know what the maximum number of bits the underlying sponge can take up is, so that they can decide how much original key material to provide. Because while it is true that the output of the XOF is indistinguishable from random with zero or 1 bits of key, giving a tiny key leaves an attacker a tiny search space.

Perhaps it's a terminology problem: reading from an XOF is like getting a hash of the input, and there size does not matter. But reading from an XOF in order to call XORKeyStream is encrypting/decrypting, and then input size matters, A LOT. This is why I called it "KeySize".

Reply:
Yes I agree for your last paragraph. So `KeySIze` means `the maximum key size you can give to the XOF,i.e. the required key size if you wanna be secure`, but the `Size` ... I'm not sure we need it in fact ? It's confusing for me at least, maybe @Daeinar can chim in. It seems we don't really need the `Size` then because we should be able to retrive any number of bytes from a XOF, so it's not clear what this size represents.

Reply:
and BTW, we need to be more consistent: what's the difference between `seed` and `key`. I create a `XOF` with a `seed` but I can retrieve its `KeySize` ? Same for in-method code.

Reply:
I added Size to help sign/anon know how big a MAC should be. Before this, it was using KeySize() to decide how much to take out of the hash, and was making signatures bigger than it needed to.

Reply:
From what you are saying it seems like it's `sign/anon` that does a bad job in the sense that it should know how many random looking bytes it has to take no ? I mean random looking bytes from sha3 or blake2 are still supposed to be random looking so regarding the security, the size of the hash to take is not dependant on the underlying hash function (as long as the hash function is correct), right ?
Is there no way to make `sign/anon` work while removing that `Size` function ?

Reply:
Sure, just hard code it to be what it was before. Doing that now.